### PR TITLE
implement builtin lpad & lpad_binary

### PR DESCRIPTION
**Pull request by niedhui**
Signed-off-by: niedhui <niedhui@gmail.com>

## What have you changed? (mandatory)

implement builtin string function: Lpad and LpadBinary

## What are the type of the changes? (mandatory)

Improvement 

## How has this PR been tested? (mandatory)

Unit Test

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 1]
```diff
@@ -650,62 +656,127 @@ impl ScalarFunc {
...
+                } else {
+                    input[..target_len].to_vec()
+                };
+                Ok(Some(Cow::Owned(r)))
+            }
+        }
+    }
+
+    #[inline]
+    fn valid_target_len_for_pad(
```
Comment (zhangjinpeng1987) [inline src/coprocessor/dag/expr/builtin_string.rs]: Pleas also add some tests for function `valid_target_len_for_pad`, thank you.

Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_string.rs]: hi @zhangjinpeng1987 this helper function is extracted  for remove duplication of the  pad-related functions， I think the tests of the for 'pad' functions is already cover it.

and I add docs to clarify what this helper function does.

Comment (zhangjinpeng1987) [reply src/coprocessor/dag/expr/builtin_string.rs]: `valid_target_len_for_pad` is a individual function, i think it's better to add unit test for it.

Comment (zhangjinpeng1987) [reply src/coprocessor/dag/expr/builtin_string.rs]: Comments show the function's expected behavior, tests show the exactly behavior, they are different.

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 2]
```diff
@@ -42,6 +42,12 @@ enum TrimDirection {
     Trailing,
 }
 
+enum TargetLen {
+    Invalid,
+    Zero,
```
Comment (ice1000) [inline src/coprocessor/dag/expr/builtin_string.rs]: Why not `Len(0)`?

Comment (ice1000) [reply src/coprocessor/dag/expr/builtin_string.rs]: So this can be replaced with `Option<usize>`.

Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_string.rs]: actually ,there is another value , when max_allowed_packet is pushed down, if target_len is greater than max_allowed_packet,  it should tell the pad function to return null with an warning.
but for now, Option is enough .

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 1 above)
Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_string.rs]: tests added, PTAL @zhangjinpeng1987

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 3]
```diff
@@ -683,32 +681,131 @@ impl ScalarFunc {
...
+                } else {
+                    input[..target_len].to_vec()
+                };
+                Ok(Some(Cow::Owned(r)))
+            }
+        }
     }
 }
 
+fn empty_str() -> Result<Option<Cow<'static, [u8]>>> {
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_string.rs]: I personally prefer not to introducing a function for this simple logic, although I'm fine with it.

**Pull request merged**